### PR TITLE
added maintainer and setElevatorTarget commands and tests

### PR DESCRIPTION
**Pull request by Gabriel-J-Young**

### CODE REVIEW

File: Competition2018/src/main/java/competition/subsystems/elevator/ElevatorSubsystem.java [hunk 1]
```diff
@@ -46,6 +48,7 @@ public ElevatorSubsystem(CommonLibFactory clf, XPropertyManager propMan) {
 		calibrationPower = propMan.createPersistentProperty("ElevatorCalibrationPower", 0.2);
 		maxHeightInInches = propMan.createPersistentProperty("Elevator Max HeightInInches", 80);
 		minHeightInInches = propMan.createPersistentProperty("Elevator Min HeightInInches", 3);		
+		targetHeight = propMan.createEphemeralProperty("targetHeight", minHeightInInches.get());
```
Comment (JohnGilb) [inline Competition2018/src/main/java/competition/subsystems/elevator/ElevatorSubsystem.java]: rename to "Elevator Target Height"

File: Competition2018/src/main/java/competition/subsystems/elevator/ElevatorSubsystem.java [hunk 2]
```diff
@@ -89,6 +92,18 @@ public void setPower(double power) {
 	    
 	    motor.simpleSet(power);
 	}
+	
+	public void maintainer() {
```
Comment (JohnGilb) [inline Competition2018/src/main/java/competition/subsystems/elevator/ElevatorSubsystem.java]: delete this method

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/MaintainerCommand.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+import xbot.common.command.BaseCommand;
+import xbot.common.math.PIDFactory;
+import competition.subsystems.elevator.ElevatorSubsystem;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import xbot.common.math.PIDFactory;
+import xbot.common.math.PIDManager;
+
+public class MaintainerCommand extends BaseCommand {
```
Comment (JohnGilb) [inline Competition2018/src/main/java/competition/subsystems/elevator/commands/MaintainerCommand.java]: `ElevatorMaintainerCommand`

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/MaintainerCommand.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    }
+    
+    @Override
+    public void initialize() {
+        log.info("Initializing");        
+    }
+
+    @Override
+    public void execute() {
+
```
Comment (JohnGilb) [inline Competition2018/src/main/java/competition/subsystems/elevator/commands/MaintainerCommand.java]: remove blank line

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/MaintainerCommand.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    @Override
+    public void initialize() {
+        log.info("Initializing");        
+    }
+
+    @Override
+    public void execute() {
+        
+        double power = pid.calculate(elevator.getTargetHeight(), elevator.currentHeight());
+
```
Comment (JohnGilb) [inline Competition2018/src/main/java/competition/subsystems/elevator/commands/MaintainerCommand.java]: remove blank line

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/MaintainerCommand.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    @Override
+    public void execute() {
+        
+        double power = pid.calculate(elevator.getTargetHeight(), elevator.currentHeight());
+       
+        elevator.setPower(power);
+        
+        if(elevator.currentHeight() >= elevator.getTargetHeight() -.01 && elevator.currentHeight() <= elevator.getTargetHeight() + .01) {
```
Comment (JohnGilb) [inline Competition2018/src/main/java/competition/subsystems/elevator/commands/MaintainerCommand.java]: This command should never finish. Remove this.

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/MaintainerCommand.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+            isFinished = true;
+        }
+        else {
+            isFinished = false;
+        }
+    }
+    
+    @Override
+    public boolean isFinished() {
```
Comment (JohnGilb) [inline Competition2018/src/main/java/competition/subsystems/elevator/commands/MaintainerCommand.java]: remove this method. We want the maintainer to run nearly 100% of the time.

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/SetElevatorTargetHeightCommand.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
+package competition.subsystems.elevator.commands;
+
+import xbot.common.command.BaseSetpointCommand;
+import xbot.common.command.SupportsSetpointLock;
+
+import com.google.inject.Inject;
+
+import competition.subsystems.elevator.ElevatorSubsystem;
+
```
Comment (JohnGilb) [inline Competition2018/src/main/java/competition/subsystems/elevator/commands/SetElevatorTargetHeightCommand.java]: remove blank line

Comment (JohnGilb) [reply Competition2018/src/main/java/competition/subsystems/elevator/commands/SetElevatorTargetHeightCommand.java]: There should only be 1 blank line between the end of the imports and the start of the class declaration

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/SetElevatorTargetHeightCommand.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        super(elevator);
+        this.elevator = elevator;
+    }
+
+    public void setGoalHeight(double height) {
+        this.height = height;
+    }
+  
+    @Override
+    public void initialize() {
```
Comment (JohnGilb) [inline Competition2018/src/main/java/competition/subsystems/elevator/commands/SetElevatorTargetHeightCommand.java:23]: add `log.info("Initializing");

File: Competition2018/src/test/java/competition/subsystems/elevator/commands/MaintainerCommandTest.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    
+    @Test
+    public void testSimple() {
+        command.initialize();
+        command.execute();
+    }
+    
+    @Test
+    public void checkGoUp() {
+        ((MockCANTalon)elevator.motor).setPosition(1200);
```
Comment (JohnGilb) [inline Competition2018/src/test/java/competition/subsystems/elevator/commands/MaintainerCommandTest.java:34]: Your test says `checkGoUp` but then you move the elevator up very high (and I'm assuming the target height is 0), and then run the command. That sounds like `checkGoDown` to me. =]

You should explicitly set the target height for these tests.

**Review by JohnGilb — CHANGES_REQUESTED** [review]

File: Competition2018/src/main/java/competition/subsystems/elevator/ElevatorSubsystem.java (hunk 1 above)
Comment (Gabriel-J-Young) [reply Competition2018/src/main/java/competition/subsystems/elevator/ElevatorSubsystem.java]: done

File: Competition2018/src/main/java/competition/subsystems/elevator/ElevatorSubsystem.java (hunk 2 above)
Comment (Gabriel-J-Young) [reply Competition2018/src/main/java/competition/subsystems/elevator/ElevatorSubsystem.java]: done

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/MaintainerCommand.java (hunk 3 above)
Comment (Gabriel-J-Young) [reply Competition2018/src/main/java/competition/subsystems/elevator/commands/MaintainerCommand.java]: done

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/MaintainerCommand.java (hunk 5 above)
Comment (Gabriel-J-Young) [reply Competition2018/src/main/java/competition/subsystems/elevator/commands/MaintainerCommand.java]: done

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/MaintainerCommand.java (hunk 6 above)
Comment (Gabriel-J-Young) [reply Competition2018/src/main/java/competition/subsystems/elevator/commands/MaintainerCommand.java]: done

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/MaintainerCommand.java (hunk 7 above)
Comment (Gabriel-J-Young) [reply Competition2018/src/main/java/competition/subsystems/elevator/commands/MaintainerCommand.java]: done

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/SetElevatorTargetHeightCommand.java (hunk 8 above)
Comment (Gabriel-J-Young) [reply Competition2018/src/main/java/competition/subsystems/elevator/commands/SetElevatorTargetHeightCommand.java]: done

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/SetElevatorTargetHeightCommand.java (hunk 9 above)
Comment (Gabriel-J-Young) [reply Competition2018/src/main/java/competition/subsystems/elevator/commands/SetElevatorTargetHeightCommand.java:23]: done

File: Competition2018/src/test/java/competition/subsystems/elevator/commands/MaintainerCommandTest.java (hunk 10 above)
Comment (Gabriel-J-Young) [reply Competition2018/src/test/java/competition/subsystems/elevator/commands/MaintainerCommandTest.java:34]: done

File: Competition2018/src/test/java/competition/subsystems/elevator/commands/MaintainerCommandTest.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        assertTrue(elevator.motor.getMotorOutputPercent() < 0);
+    }
+    
+    @Test
+    public void Stop() {
+        ((MockCANTalon)elevator.motor).setPosition(7000);
+        elevator.setTargetHeight(70);
+        command.initialize();
+        command.execute();
+        assertTrue(elevator.motor.getMotorOutputPercent() <= .01);
```
Comment (JohnGilb) [inline Competition2018/src/test/java/competition/subsystems/elevator/commands/MaintainerCommandTest.java:56]: make this `assertEquals(0, elevator.motor.getMotorOutputPercent(), 0.001);`

**Review by JohnGilb — APPROVED** [review]

**Pull request merged**